[sharded_tensor] add readonly tensor properties

wanchaol opened the pull request:
Summary: This PR add read only tensor properties to sharded tensor, to match the torch.Tensor behaviors.

Test Plan: test_sharded_tensor_metadata

Differential Revision: D30459343

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -590,6 +590,18 @@ def size(self) -> torch.Size:
         """
         return self._metadata.size
 
+    @property
+    def requires_grad(self):
+        return self._metadata.requires_grad
```
nit: should we add the rest as well `pin_memory`, `memory_layout` etc.? Also, should `size` be `property` too?

wanchaol replied:
good point, I can add `is_pinned()` although that's not a property. I looked into tensor attributes page https://pytorch.org/docs/stable/tensor_attributes.html and tried locally, the local tensor doesn't have `.memory_layout` properties exposed, maybe we should only expose a `is_contiguous()`.

For `size`, iirc for local tensors, user can call `size()`, but couldn't do `.size`, instead they do `.shape`, so we might expose `shape` if we want?

Commit:
[sharded_tensor] add readonly tensor properties (#63679)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/63679

This PR add read only tensor properties to sharded tensor, to match the torch.Tensor behaviors.

Test Plan: test_sharded_tensor_metadata

Reviewed By: pritamdamania87

Differential Revision: D30459343

fbshipit-source-id: 38d167b03abb1306c00275fd12d76504cc3a553f
